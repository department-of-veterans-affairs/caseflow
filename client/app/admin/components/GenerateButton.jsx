/* eslint-disable func-style */
import React, { useEffect, useState } from 'react';
import Button from '../../components/Button';
import Modal from '../../components/Modal';
import Alert from '../../components/Alert';
import LoadingContainer from '../../components/LoadingContainer';
import { LOGO_COLORS } from 'app/constants/AppConstants';
import PropTypes from 'prop-types';
import { CSVLink } from 'react-csv';

const GenerateButton = (props) => {

  // state properties
  const [showConfirmModal, setShowConfirmModal] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [showEmptyResultsBanner, setShowEmptyResultsBanner] = useState(false);
  const [showErrorBanner, setShowErrorBanner] = useState(false);
  const [extractedResults, setExtractedResults] = useState('');

  useEffect(() => {
    setIsLoading(props.isLoading);
  }, [props.isLoading]);

  useEffect(() => {
    setExtractedResults(props.extractedResults);

    if (props.extractedResults?.length > 0) {
      setShowEmptyResultsBanner(false);
      setShowConfirmModal(true);
    }
  }, [props.extractedResults]);

  useEffect(() => {
    if (props.emptyResultsMessage?.length > 0) {
      setShowEmptyResultsBanner(true);
      setShowConfirmModal(false);
    }
  }, [props.emptyResultsMessage]);

  useEffect(() => {
    if (props.manualExtractionSuccess === false) {
      setShowErrorBanner(true);
    } else {
      setShowErrorBanner(false);
    }
  }, [props.manualExtractionSuccess]);

  const onClickGenerate = () => {
    props.sendExtractRequest();
  };

  return (
    <div style={{ height: '75vh' }}>
      {
        showEmptyResultsBanner &&
        <div style={{ padding: '10px' }}>
          <Alert message="No Veterans were found" type="success" />
        </div>
      }
      {
        showErrorBanner &&
        <div style={{ padding: '10px' }}>
          <Alert message="Veteran Extract Failed" type="error" />
        </div>
      }
      {
        !isLoading &&
        <Button
          id="generate-extract"
          onClick={() => onClickGenerate()}
        >
          Generate Veteran Extract
        </Button>
      }
      {
        showConfirmModal &&

        <Modal title="This file contains PII"
          confirmButton={<CSVLink data={extractedResults} filename="veteran_extract.csv" onClick={() => setShowConfirmModal(false)}>Confirm</CSVLink>}
          closeHandler={() => {
            setShowConfirmModal(false);
          }}
        >
          Are you sure you want to download it?
        </Modal>
      }
      {
        isLoading &&

        <LoadingContainer color={LOGO_COLORS.QUEUE.ACCENT}>
          <div className="loading-div">
            Action is Running...
          </div>
        </LoadingContainer>
      }
    </div>
  );
};

GenerateButton.propTypes = {
  extractedResults: PropTypes.string,
  manualExtractionSuccess: PropTypes.bool,
  isLoading: PropTypes.bool,
  sendExtractRequest: PropTypes.func,
  emptyResultsMessage: PropTypes.string,
};

export default GenerateButton;
