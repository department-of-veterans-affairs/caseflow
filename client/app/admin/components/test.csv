VET_LAST_NAME,VET_FIRST_NAME,VET_MIDDLE_NAME,VET_DATE_OF_BIRTH,VET_SSN,VET_PARTICIPANT_ID,VET_FILE_NUMBER,APPELLANT_LAST_NAME,APPELLANT_FIRST_NAME,APPELLANT_MIDDLE_NAME,APPELLANT_DATE_OF_BIRTH,APPELLANT_SSN,APPELLANT_GENDER,APPELLANT_ADDRESS,APPELLANT_PHONE,APPELLANT_EMAIL,APPELLANT_EDI_PI,APPELLANT_CORP_PID,APPELLANT_VACOLS_INTERNAL_ID,RELATIONSHIP_TO_VETERAN