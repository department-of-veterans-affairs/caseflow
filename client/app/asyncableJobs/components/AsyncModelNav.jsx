import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import DropdownButton from '../../components/DropdownButton';

const DATE_TIME_FORMAT = 'ddd MMM DD HH:mm:ss YYYY';

export default class AsyncModelNav extends React.PureComponent {
  modelNameLinks = () => {
    const links = [];
    const modelNames = this.props.models.sort();
    const numLinks = modelNames.length;

    for (let modelName of modelNames) {
      const url = `/asyncable_jobs/${modelName}/jobs`;
      let modelLink;

      if (numLinks > 4) {
        modelLink = {
          title: modelName,
          target: url
        };
      } else {
        modelLink = <span key={modelName} className="cf-model-jobs-link"><a href={url}>{modelName}</a></span>;
      }

      links.push(modelLink);
    }

    return numLinks > 4 ? <DropdownButton lists={links} label="Filter by Job Type" /> : links;
  }

  render = () => {
    const currentPath = this.props.asyncableJobKlass ?
      `/asyncable_jobs/${this.props.asyncableJobKlass}/jobs` :
      '/jobs';

    return <div>
      <strong>Last updated:</strong> {moment(this.props.fetchedAt).format(DATE_TIME_FORMAT)}
      <div style={{ marginTop: '.5em' }}>
        <a style={{ marginRight: '.5em' }} href="/jobs" className="cf-link-btn">All jobs</a>
        {this.modelNameLinks()}
        <a style={{ float: 'right' }} href={`${currentPath}.csv`} className="cf-link-btn">Download as CSV</a>
      </div>
    </div>;
  }
}

AsyncModelNav.propTypes = {
  models: PropTypes.array,
  fetchedAt: PropTypes.string,
  asyncableJobKlass: PropTypes.string
};
