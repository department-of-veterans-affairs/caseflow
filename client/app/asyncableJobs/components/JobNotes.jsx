import React from 'react';
import Button from '../../components/Button';
import ApiUtil from '../../util/ApiUtil';
import PropTypes from 'prop-types';
import moment from 'moment';
import ReactMarkdown from 'react-markdown';

const DATE_TIME_FORMAT = 'ddd MMM DD HH:mm:ss YYYY';

import Checkbox from '../../components/Checkbox';
import TextareaField from '../../components/TextareaField';

class NewNoteForm extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      saveInProgress: false,
      checkboxSelected: true,
      newNote: ''
    };
  }

  onSave = () => {
    const job = this.props.job;
    const url = `/asyncable_jobs/${job.klass}/jobs/${job.id}/note`;
    const form = this;
    const notes = this.props.notes;

    const data = {
      note: this.state.newNote,
      send_to_intake_user: this.state.checkboxSelected
    };

    notes.setState({ reloaded: false });

    ApiUtil.post(url, { data }).
      then(
        (response) => {
          const newNote = JSON.parse(response.text);

          form.setState({
            saveInProgress: false,
            checkboxSelected: false,
            newNote: ''
          });

          // add note to current page.
          notes.props.notes.unshift(newNote);
          notes.setState({ reloaded: true });
        }
      );
  }

  onChange = (event) => {
    this.setState({ newNote: event });
  }

  onCheckboxChange = (event) => {
    this.setState({ checkboxSelected: event });
  }

  isSaveDisabled = () => {
    return this.state.saveInProgress || this.state.newNote.trim().length === 0;
  }

  isCheckboxChecked = () => {
    return this.state.checkboxSelected;
  }

  sendToUserCheckbox = () => {
    return <Checkbox
      label="Send as message to user"
      name="send_to_intake_user"
      value={this.isCheckboxChecked()}
      onChange={this.onCheckboxChange}
      disabled={this.state.saveInProgress}
    />;
  }

  render = () => {
    return <div className="comment-size-container">
      <TextareaField
        hideLabel
        name="Add Note"
        aria-label="Add Note"
        onChange={this.onChange}
        value={this.state.newNote}
      />
      <div className="comment-save-button-container">
        <span className="cf-right-side">
          { this.props.job.user && this.sendToUserCheckbox() }
          <Button
            name="save"
            disabled={this.isSaveDisabled()}
            onClick={this.onSave}>
              Add Note
          </Button>
        </span>
      </div>
    </div>;
  }

}

NewNoteForm.propTypes = {
  job: PropTypes.object,
  notes: PropTypes.object,
  onSave: PropTypes.func,
  onCancel: PropTypes.func
};

export default class JobNotes extends React.PureComponent {

  render = () => {
    const { notes, job } = this.props;

    return <div className="job-notes">
      <h3>Notes</h3>
      <NewNoteForm job={job} notes={this} />
      <div>
        { notes.map((note, index) => {
          return <div className="job-note" key={`job-note-container-${index}`}>
            <div
              className="job-note-details"
              data-key={`job-note-${index}`}
              key={`job-note-${index}`}
              id={`job-note-${note.id}`}>

              <div>
                <span className="job-note-time">{moment(note.created_at).format(DATE_TIME_FORMAT)}</span>
                <span className="job-note-user">{note.user}</span>
              </div>
              <div className="job-note-note"><ReactMarkdown source={note.note} /></div>

            </div>
          </div>;
        })}
      </div>
    </div>;
  }
}

JobNotes.propTypes = {
  job: PropTypes.object,
  notes: PropTypes.array
};
