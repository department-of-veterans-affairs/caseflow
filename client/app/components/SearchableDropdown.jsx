import React, { Component, PropTypes } from 'react';
import Select from 'react-select';

class SearchableDropdown extends Component {

  constructor(props) {
    super(props);
    this.state = {
      value: null
    };
  }

  onChange = (value) => {
    let newValue = value;

    /*
     * this is a temp fix for react-select value backspace
     * issue.
     * Setting value to null when an option is deselected
     * using the backspace.
     * https://github.com/JedWatson/react-select/pull/773
     */

    if (Array.isArray(value) && value.length <= 0) {
      newValue = null;
    }
    this.setState({ value: newValue });
    if (this.props.onChange) {
      this.props.onChange(newValue);
    }
  }

  render() {
    const {
      options,
      placeholder,
      errorMessage,
      label,
      name,
      required,
      readOnly
    } = this.props;

    return <div className="cf-form-dropdown">
      <label className="question-label" htmlFor={name}>
        {label || name} {required && <span className="cf-required">Required</span>}
      </label>
      {errorMessage && <span className="usa-input-error-message">{errorMessage}</span>}
      <Select
        id={name}
        value={this.state.value}
        options={options}
        onChange={this.onChange}
        placeholder={placeholder ? placeholder : "Select option"}
        clearable={false}
        noResultsText="Not an option"
        disabled={readOnly}
      />
    </div>;
  }
}

SearchableDropdown.propTypes = {
  errorMessage: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  options: PropTypes.array,
  readOnly: PropTypes.bool,
  required: PropTypes.bool,
  placeholder: PropTypes.string
};

export default SearchableDropdown;
