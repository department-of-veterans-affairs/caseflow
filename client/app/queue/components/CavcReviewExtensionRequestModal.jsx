import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { startCase } from 'lodash';
import { sprintf } from 'sprintf-js';

import Modal from '../../components/Modal';
import RadioField from '../../components/RadioField';
import SearchableDropdown from '../../components/SearchableDropdown';
import TextareaField from '../../components/TextareaField';
import TextField from '../../components/TextField';

import ApiUtil from '../../util/ApiUtil';
import { showSuccessMessage } from '../uiReducer/uiActions';

import { COLOCATED_HOLD_DURATIONS, CUSTOM_HOLD_DURATION_TEXT } from '../constants';
import COPY from '../../../COPY';
import Alert from '../../components/Alert';

const decisions = [
  COPY.CAVC_EXTENSION_REQUEST_GRANT,
  COPY.CAVC_EXTENSION_REQUEST_DENY
];
const helpText = {
  [COPY.CAVC_EXTENSION_REQUEST_GRANT]: COPY.CAVC_EXTENSION_REQUEST_GRANT_HELP_TEXT,
  [COPY.CAVC_EXTENSION_REQUEST_DENY]: COPY.CAVC_EXTENSION_REQUEST_DENY_HELP_TEXT
};
const decisionOptions = decisions.map((value) => (
  {
    value,
    displayText: startCase(value),
    help: helpText[value]
  }
));

/**
 * Modal to allow a user to grant or deny a cavc remand extension request. All fields are required.
 */
export const CavcReviewExtensionRequestModalUnconnected = ({ onCancel, onSubmit, error }) => {
  const [decision, setDecision] = useState();
  const [holdDuration, setHoldDuration] = useState();
  const [customHoldDuration, setCustomHoldDuration] = useState();
  const [instructions, setInstructions] = useState();
  const [highlightFormItems, setHighlightFormItems] = useState(false);

  const granted = () => decision === decisions[0];
  const usingCustomHold = () => holdDuration?.value === CUSTOM_HOLD_DURATION_TEXT;

  const validDecision = () => Boolean(decision);
  // Don't need to validate on hold duration if going deny route
  const validHoldDuration = () => !granted() || Boolean(holdDuration?.value);
  // Don't need to validate custom on hold duration if going deny route or if using a pre-selected hold value
  const validCustomHoldDuration = () => !granted() || !usingCustomHold() || customHoldDuration > 0;
  const validInstructions = () => Boolean(instructions);
  const validateForm = () => validDecision() && validHoldDuration() && validCustomHoldDuration() && validInstructions();

  const cancel = () => onCancel();
  const submit = () => {
    if (validateForm()) {
      const selectedHoldDuration = usingCustomHold() ? customHoldDuration : holdDuration?.value;

      onSubmit(decision, instructions, granted() ? selectedHoldDuration : null);
    } else {
      setHighlightFormItems(true);
    }
  };

  const modalButtons = [
    {
      classNames: ['cf-modal-link', 'cf-btn-link'],
      name: 'Cancel',
      onClick: cancel,
    },
    {
      classNames: ['usa-button', 'usa-button-secondary'],
      name: 'Confirm',
      onClick: submit,
    },
  ];

  const decisionField = <RadioField
    name="decision"
    label={COPY.CAVC_EXTENSION_REQUEST_DECISION_LABEL}
    errorMessage={highlightFormItems && !validDecision() ? 'Choose one' : null}
    value={decision}
    onChange={setDecision}
    options={decisionOptions}
    strongLabel
  />;

  const holdDurationField = () => (
    granted() ? <SearchableDropdown
      name="duration"
      label={COPY.COLOCATED_ACTION_PLACE_HOLD_LENGTH_SELECTOR_LABEL}
      placeholder={COPY.COLOCATED_ACTION_PLACE_HOLD_LENGTH_SELECTOR_LABEL}
      errorMessage={highlightFormItems && !validHoldDuration() ? 'Choose one' : null}
      value={holdDuration}
      onChange={setHoldDuration}
      options={COLOCATED_HOLD_DURATIONS.map((value) => ({
        label: Number(value) ? `${value} days` : value,
        value
      }))}
    /> : null
  );

  const customHoldField = () => (
    granted() && usingCustomHold() ? <TextField
      name="customDuration"
      label={COPY.COLOCATED_ACTION_PLACE_CUSTOM_HOLD_COPY}
      type="number"
      value={customHoldDuration}
      onChange={setCustomHoldDuration}
      errorMessage={highlightFormItems && !validCustomHoldDuration() ?
        COPY.COLOCATED_ACTION_PLACE_CUSTOM_HOLD_INVALID_VALUE : null
      }
      inputProps={{ min: 1 }}
    /> : null
  );

  const instructionsField = <TextareaField
    name="instructions"
    label={COPY.PROVIDE_INSTRUCTIONS_AND_CONTEXT_LABEL}
    errorMessage={highlightFormItems && !validInstructions() ? COPY.CAVC_INSTRUCTIONS_ERROR : null}
    value={instructions}
    onChange={setInstructions}
  />;

  const errorAlert = () => error?.title ? <Alert title={error.title} message={error?.detail} type="error" /> : null;

  return (
    <Modal
      title={COPY.CAVC_EXTENSION_REQUEST_TITLE}
      buttons={modalButtons}
      closeHandler={onCancel}
    >
      { errorAlert() }
      { decisionField }
      { holdDurationField() }
      { customHoldField() }
      { instructionsField }
    </Modal>
  );
};

CavcReviewExtensionRequestModalUnconnected.propTypes = {

  /**
   * Callback for when the user presses the confirm button and the form is valid. Will be passed the following params
   *
   * @param {string} decision     The decision the user has made, one of "grant" or "deny"
   * @param {string} instructions The instructions provided with the decision
   * @param {number} holdDuration The number of days to grant an extension for. Will be null if the extension request is
   *                              denied
   */
  onSubmit: PropTypes.func,

  /**
   * Callback for when the user presses the cancel button
   */
  onCancel: PropTypes.func,

  /**
   * Any errors to display in the modal. Automatically set if the request was not successful
   */
  error: PropTypes.shape({
    title: PropTypes.string,
    detail: PropTypes.string
  })
};

const CavcReviewExtensionRequestModal = ({ closeModal, taskId }) => {
  const dispatch = useDispatch();
  const [error, setError] = useState();

  const onCancel = () => closeModal();
  const onSubmit = (decision, instructions, holdDuration) => {
    const payload = {
      data: {
        task: {
          days_on_hold: holdDuration,
          instructions: [instructions],
          decision
        }
      }
    };

    ApiUtil.post(`/tasks/${taskId}/extension_request/`, payload).then(() => {
      const successMessage = decision === 'grant' ?
        {
          title: sprintf(COPY.CAVC_EXTENSION_REQUEST_GRANT_SUCCESS_TITLE, holdDuration),
          detail: COPY.CAVC_EXTENSION_REQUEST_GRANT_SUCCESS_DETAIL
        } : {
          title: COPY.CAVC_EXTENSION_REQUEST_DENY_SUCCESS_TITLE,
          detail: COPY.CAVC_EXTENSION_REQUEST_DENY_SUCCESS_DETAIL
        };

      dispatch(showSuccessMessage(successMessage));
      closeModal();
    }, (err) => {
      setError(err.response.body.errors[0]);
    });
  };

  return <CavcReviewExtensionRequestModalUnconnected
    onCancel={onCancel}
    onSubmit={onSubmit}
    error={error}
  />;
};

CavcReviewExtensionRequestModal.propTypes = {

  /**
   * Callback to close the modal and return to the previous page
   */
  closeModal: PropTypes.func,

  /**
   * The id of the CavcRemandProcessedLetterResponseWindowTask the extension request is for
   */
  taskId: PropTypes.string
};

export default CavcReviewExtensionRequestModal;
