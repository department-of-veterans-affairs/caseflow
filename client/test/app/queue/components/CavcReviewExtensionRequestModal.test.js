import React from 'react';
import { mount } from 'enzyme';

import { CavcReviewExtensionRequestModalUnconnected } from 'app/queue/components/CavcReviewExtensionRequestModal';
import { SearchableDropdown } from 'app/components/SearchableDropdown';

import COPY from 'COPY';

describe('CavcReviewExtensionRequestModal', () => {
  beforeEach(() => jest.clearAllMocks());

  const onSubmit = jest.fn();
  const onCancel = jest.fn();

  const setup = (args = {}) => {
    return mount(
      <CavcReviewExtensionRequestModalUnconnected
        onCancel={onCancel}
        onSubmit={onSubmit}
        {...args}
      />
    );
  };

  const clickSubmit = (modal) => modal.find('button#Review-extension-request-button-id-1').simulate('click');
  const clickCancel = (modal) => modal.find('button#Review-extension-request-button-id-0').simulate('click');
  const selectGrant = (modal) => modal.find('#decision_grant').simulate('change', { target: { value: 'grant' } });
  const selectDeny = (modal) => modal.find('#decision_deny').simulate('change', { target: { value: 'deny' } });
  const select15DayDuration = (modal) => {
    const dropdown = modal.find(SearchableDropdown);

    // Select the first option ("15 days") and press enter
    dropdown.find('Select').simulate('keyDown', { key: 'ArrowDown', keyCode: 40 });
    dropdown.find('Select').simulate('keyDown', { key: 'Enter', keyCode: 13 });
  };
  const selectCustomDuration = (modal) => {
    const dropdown = modal.find(SearchableDropdown);

    // Select the last option ("custom") and press enter
    dropdown.find('Select').simulate('keyDown', { key: 'ArrowUp', keyCode: 38 });
    dropdown.find('Select').simulate('keyDown', { key: 'Enter', keyCode: 13 });
  };
  const populateCustomDuration = (modal, duration) => (
    modal.find('#customDuration').simulate('change', { target: { value: duration } })
  );
  const populateInstructions = (modal, instructions) => (
    modal.find('#instructions').simulate('change', { target: { value: instructions } })
  );

  it('renders correctly', async () => {
    const extensionModal = setup();

    expect(extensionModal).toMatchSnapshot();
  });

  it('shows hold duration selector only when decision is grant', () => {
    const extensionModal = setup();

    expect(extensionModal.find('input#duration').length).toBe(0);
    selectGrant(extensionModal);
    expect(extensionModal.find('input#duration').length).toBe(1);
    selectDeny(extensionModal);
    expect(extensionModal.find('input#duration').length).toBe(0);
  });

  it('shows custom hold duration selector when only decision is grant and duration is custom', () => {
    const extensionModal = setup();

    expect(extensionModal.find('input#customDuration').length).toBe(0);
    selectGrant(extensionModal);
    expect(extensionModal.find('input#customDuration').length).toBe(0);
    select15DayDuration(extensionModal);
    expect(extensionModal.find('input#customDuration').length).toBe(0);
    selectCustomDuration(extensionModal);
    expect(extensionModal.find('input#customDuration').length).toBe(1);
    select15DayDuration(extensionModal);
    expect(extensionModal.find('input#customDuration').length).toBe(0);
  });

  it('displays an error if provided', () => {
    const title = 'Error title';
    const detail = 'Error message';
    const extensionModal = setup({ error: { title, detail } });

    expect(extensionModal.find('.usa-alert-error').length).toBe(1);
    expect(extensionModal.find('.usa-alert-heading').props().children).toBe(title);
    expect(extensionModal.find('.usa-alert-text').props().children).toBe(detail);
  });

  it('calls onSubmit with selected values when "Confirm" is pressed and form is valid', () => {
    const extensionModal = setup();

    selectDeny(extensionModal);
    populateInstructions(extensionModal, 'instructions');
    clickSubmit(extensionModal);

    expect(onSubmit).toHaveBeenCalledWith('deny', 'instructions', null);

    selectGrant(extensionModal);
    select15DayDuration(extensionModal);
    populateInstructions(extensionModal, 'new instructions');
    clickSubmit(extensionModal);

    expect(onSubmit).toHaveBeenCalledWith('grant', 'new instructions', 15);

    selectCustomDuration(extensionModal);
    populateInstructions(extensionModal, 'new new instructions');
    populateCustomDuration(extensionModal, 25);
    clickSubmit(extensionModal);

    expect(onSubmit).toHaveBeenCalledWith('grant', 'new new instructions', 25);
  });

  it('calls onCancel ', () => {
    const extensionModal = setup();

    clickCancel(extensionModal);

    expect(onCancel).toHaveBeenCalled();
  });

  describe('form validations', () => {
    const errorClass = '.usa-input-error-message';

    const validationErrorShows = (modal, errorMessage) => {
      clickSubmit(modal);

      return modal.find(errorClass).findWhere((node) => node.props().children === errorMessage).length > 0;
    };

    describe('decision type validations', () => {
      const error = 'Choose one';

      it('shows error on no decision type selection', () => {
        const extensionModal = setup();

        expect(validationErrorShows(extensionModal, error)).toBeTruthy();
      });

      it('does not show error on correctly selected decision type', () => {
        const extensionModal = setup();

        selectDeny(extensionModal);

        expect(validationErrorShows(extensionModal, error)).toBeFalsy();
      });
    });

    describe('on hold duration validations', () => {
      const error = 'Choose one';

      it('shows error on no selected on hold duration', () => {
        const extensionModal = setup();

        selectGrant(extensionModal);

        expect(validationErrorShows(extensionModal, error)).toBeTruthy();
      });

      it('does not show error on selected on hold duration', () => {
        const extensionModal = setup();

        selectGrant(extensionModal);
        select15DayDuration(extensionModal);

        expect(validationErrorShows(extensionModal, error)).toBeFalsy();
      });
    });

    describe('custom on hold duration validations', () => {
      const error = COPY.COLOCATED_ACTION_PLACE_CUSTOM_HOLD_INVALID_VALUE;

      it('shows error on no selected custom hold duration', () => {
        const extensionModal = setup();

        selectGrant(extensionModal);
        selectCustomDuration(extensionModal);

        expect(validationErrorShows(extensionModal, error)).toBeTruthy();
      });

      it('shows error on selected duration less than 1', () => {
        const extensionModal = setup();

        selectGrant(extensionModal);
        selectCustomDuration(extensionModal);
        populateCustomDuration(extensionModal, -1);

        expect(validationErrorShows(extensionModal, error)).toBeTruthy();
      });

      it('does not show error on valid custom hold duration', () => {
        const extensionModal = setup();

        selectGrant(extensionModal);
        selectCustomDuration(extensionModal);
        populateCustomDuration(extensionModal, 25);

        expect(validationErrorShows(extensionModal, error)).toBeFalsy();
      });
    });

    describe('instruction validations', () => {
      const error = COPY.CAVC_INSTRUCTIONS_ERROR;

      it('shows error on no provided instructions', () => {
        const extensionModal = setup();

        expect(validationErrorShows(extensionModal, error)).toBeTruthy();
      });

      it('does not show error on provided instructions', () => {
        const extensionModal = setup();

        populateInstructions(extensionModal, 'here are some instructions');

        expect(validationErrorShows(extensionModal, error)).toBeFalsy();
      });
    });
  });
});
