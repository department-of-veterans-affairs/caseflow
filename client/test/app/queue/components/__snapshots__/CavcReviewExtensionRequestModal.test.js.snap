// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CavcReviewExtensionRequestModal renders correctly 1`] = `
<CavcReviewExtensionRequestModalUnconnected
  onCancel={[MockFunction]}
  onSubmit={[MockFunction]}
>
  <Modal
    buttons={
      Array [
        Object {
          "classNames": Array [
            "cf-modal-link",
            "cf-btn-link",
          ],
          "name": "Cancel",
          "onClick": [Function],
        },
        Object {
          "classNames": Array [
            "usa-button",
            "usa-button-secondary",
          ],
          "name": "Confirm",
          "onClick": [Function],
        },
      ]
    }
    className=""
    closeHandler={[MockFunction]}
    scrollLock={true}
    title="Review extension request"
  >
    <section
      aria-describedby="modal_id-desc"
      aria-labelledby="modal_id-title"
      aria-modal="true"
      className="cf-modal active "
      id="modal_id"
      role="dialog"
    >
      <TouchProvider>
        <ScrollLock
          accountForScrollbars={true}
        >
          <Sheet
            styles="body {
      box-sizing: border-box !important;
      overflow: hidden !important;
      position: relative !important;
      
      padding-right: 1024px !important;
    }"
          />
        </ScrollLock>
      </TouchProvider>
      <div
        className="cf-modal-body"
        id=""
      >
        <button
          className="cf-modal-close"
          id="Review-extension-request-button-id-close"
          onClick={[MockFunction]}
          type="button"
        >
          <span
            className="usa-sr-only"
          >
            Close
          </span>
          <CloseIcon
            className="cf-icon-close"
            color="#212121"
            size={55}
          >
            <svg
              className="cf-icon-close"
              height={55}
              viewBox="0 0 55 55"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                close
              </title>
              <g
                fill="#212121"
              >
                <path
                  d="M52.6 46.9l-6 6c-.8.8-1.9 1.2-3 1.2s-2.2-.4-3-1.2l-13-13-13 13c-.8.8-1.9 1.2-3 1.2s-2.2-.4-3-1.2l-6-6c-.8-.8-1.2-1.9-1.2-3s.4-2.2 1.2-3l13-13-13-13c-.8-.8-1.2-1.9-1.2-3s.4-2.2 1.2-3l6-6c.8-.8 1.9-1.2 3-1.2s2.2.4 3 1.2l13 13 13-13c.8-.8 1.9-1.2 3-1.2s2.2.4 3 1.2l6 6c.8.8 1.2 1.9 1.2 3s-.4 2.2-1.2 3l-13 13 13 13c.8.8 1.2 1.9 1.2 3s-.4 2.2-1.2 3z"
                />
              </g>
            </svg>
          </CloseIcon>
        </button>
        <div
          style={
            Object {
              "display": "flex",
            }
          }
        >
          <div
            data-css-1gs0ko2=""
          >
            <h1
              id="modal_id-title"
            >
              Review extension request
            </h1>
            <div
              data-css-o98ubo=""
            >
              <RadioField
                className={
                  Array [
                    "usa-fieldset-inputs",
                  ]
                }
                errorMessage={null}
                label="How will you proceed?"
                name="decision"
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "displayText": "Grant",
                      "help": "Task will go on hold for selected number of days",
                      "value": "grant",
                    },
                    Object {
                      "displayText": "Deny",
                      "help": "Marks the extension request as denied",
                      "value": "deny",
                    },
                  ]
                }
                required={false}
                strongLabel={true}
              >
                <fieldset
                  className="usa-fieldset-inputs cf-form-radio-inline "
                >
                  <legend
                    className=""
                  >
                    <strong>
                      <span>
                        How will you proceed?
                         
                      </span>
                    </strong>
                  </legend>
                  <div
                    className="cf-form-radio-options"
                  >
                    <div
                      className="cf-form-radio-option"
                      key="decision-grant-0"
                    >
                      <input
                        disabled={false}
                        id="decision_grant"
                        name="decision"
                        onChange={[Function]}
                        type="radio"
                        value="grant"
                      />
                      <label
                        className=""
                        htmlFor="decision_grant"
                      >
                        Grant
                      </label>
                      <RadioFieldHelpText
                        help="Task will go on hold for selected number of days"
                      >
                        <div
                          className="cf-form-radio-help helpText"
                        >
                          Task will go on hold for selected number of days
                        </div>
                      </RadioFieldHelpText>
                    </div>
                    <div
                      className="cf-form-radio-option"
                      key="decision-deny-1"
                    >
                      <input
                        disabled={false}
                        id="decision_deny"
                        name="decision"
                        onChange={[Function]}
                        type="radio"
                        value="deny"
                      />
                      <label
                        className=""
                        htmlFor="decision_deny"
                      >
                        Deny
                      </label>
                      <RadioFieldHelpText
                        help="Marks the extension request as denied"
                      >
                        <div
                          className="cf-form-radio-help helpText"
                        >
                          Marks the extension request as denied
                        </div>
                      </RadioFieldHelpText>
                    </div>
                  </div>
                </fieldset>
              </RadioField>
              <TextareaField
                characterLimitTopRight={false}
                disabled={false}
                errorMessage={null}
                label="Provide instructions and context for this action"
                name="instructions"
                onChange={[Function]}
                optional={false}
                required={false}
              >
                <div
                  className="cf-form-textarea"
                >
                  <label
                    className="question-label"
                    htmlFor="instructions"
                  >
                    <FormLabel
                      label="Provide instructions and context for this action"
                      name="instructions"
                      optional={false}
                      required={false}
                    >
                      <span>
                        Provide instructions and context for this action
                      </span>
                    </FormLabel>
                  </label>
                  <textarea
                    disabled={false}
                    id="instructions"
                    name="instructions"
                    onChange={[Function]}
                  />
                </div>
              </TextareaField>
            </div>
          </div>
        </div>
        <div
          className="cf-modal-divider"
        />
        <div
          className="cf-modal-controls"
        >
          <Button
            classNames={
              Array [
                "cf-modal-link",
                "cf-btn-link",
                "cf-push-left",
              ]
            }
            id="Review-extension-request-button-id-0"
            key="0"
            linkStyling={false}
            loading={false}
            name="Cancel"
            onClick={[Function]}
            type="button"
            willNeverBeLoading={false}
          >
            <span>
              <button
                className="cf-modal-link cf-btn-link cf-push-left usa-button"
                id="Review-extension-request-button-id-0"
                onClick={[Function]}
                type="button"
              >
                Cancel
              </button>
            </span>
          </Button>
          <Button
            classNames={
              Array [
                "usa-button",
                "usa-button-secondary",
                "cf-push-right",
              ]
            }
            id="Review-extension-request-button-id-1"
            key="1"
            linkStyling={false}
            loading={false}
            name="Confirm"
            onClick={[Function]}
            type="button"
            willNeverBeLoading={false}
          >
            <span>
              <button
                className="usa-button usa-button-secondary cf-push-right usa-button"
                id="Review-extension-request-button-id-1"
                onClick={[Function]}
                type="button"
              >
                Confirm
              </button>
            </span>
          </Button>
        </div>
      </div>
    </section>
  </Modal>
</CavcReviewExtensionRequestModalUnconnected>
`;
